### DISCUSSION TITLE
Working around method taking precedence over class of the same name

### DISCUSSION THREAD

**Discussion by Bartleby2718** [dotnet/csharplang#8216] [Q&A]
### Background knowledge
[NUnit has an abstract class `Has` with static properties like `All` or `Some`.](https://docs.nunit.org/api/NUnit.Framework.Has.html) Sample NUnit test code:
```cs
Assert.That(new[] { "a", "b" }, Has.All.Match("myRegex"));
```

### My Setup
All my test classes extend the following class from an internal NuGet package.
```cs
public abstract class MyBaseTest
{
    ...
    public bool Has<T>(object? key = null) { ... }
    public bool Has(Type type, object? key = null) { ... }
    ...
}
```

### Problem
I get CS0119 if I try to add `Assert.That(new[] { "a", "b" }, Has.All.Match("myRegex"));` to my test class:
> 'MyBaseTest.Has<T>(object?)' is a method, which is not valid in the given context

### Possible options:
1. Do not extend `MyBaseTest`
- This is not a viable option for us, as it has a lot of handy features.
2. Rename or remove the `Has` methods in `MyBaseTest`
- This is something we can consider, but I'm wondering if there are any other options.
3. Fully qualify `Has` as `NUnit.Framework`
- This defeats the purpose of using a fluent syntax in the first place.
4. Use an alias for the `Has` class (e.g. `using DoesHave = NUnit.Framework.Has;`)
- This probably won't work well with https://github.com/nunit/nunit.analyzers. We'll have to manually add the alias and rename the occurrences if the analyzers add `Has`.

### Question
Is there a feature in C# that will help me work around this problem?

**Comment by huoyaoyuan** [comment]
> 3\. Fully qualify `Has` as `NUnit.Framework`

> 4\. Use an alias for the `Has` class (e.g. `using DoesHave = NUnit.Framework.Has;`)

These are the features for this situation. There's always a chance to have name conflict in two components. And basically, the one from "near" component does win.

Note that you can use the alias like `using Has = NUnit.Framework.Has;` It will be preferred and you will need to qualify the method as `this.Has`.

**Reply by Bartleby2718** [reply]
I tried `using Has = NUnit.Framework.Has;`, but I still get the same CS0119. :(

**Reply by colejohnson66** [reply]
A using alias needs a different name than what it conflicts with. It doesn’t change priority. @huoyaoyuan’s suggestion *renames* it to `DoesHave` which would no longer conflict.

**Comment by iam3yal** [comment]
@Bartleby2718 

> Rename or remove the Has methods in MyBaseTest

In my opinion, just rename it to something else, e.g., `ThisHas`, `ItHas`, `_Has`.

**Reply by Bartleby2718** [reply]
This is definitely an option, but I wanted to minimize breaking changes because there are multiple projects (in different repositories) that use this `public` API.
